### PR TITLE
Add OnBehalfOfCredential

### LINKED ISSUE #19308: Feature: Add On-Behalf-Of (OBO) Auth Flow for the Microsoft Graph Team
The [OAuth 2.0 On-Behalf-Of flow (OBO)](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow) serves the use case where an application invokes a service/web API, which in turn needs to call another service/web API. The idea is to propagate the delegated user identity and permissions through the request chain. For the middle-tier service to make authenticated requests to the downstream service, it needs to secure an access token from the Microsoft identity platform, on behalf of the user.

**Related Links:**
Prototype: https://github.com/jongio/azidext/pull/41
Documenation: https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow
.NET Design: https://gist.github.com/christothes/15b9903f7da0954aee166166a8fa1919
MS Graph Team's tracking issue: https://github.com/microsoftgraph/msgraph-sdk-java-core/issues/132

### PR DESCRIPTION
**Pull request by chlowell**
Closes #19308

### CODE REVIEW

File: sdk/identity/azure-identity/azure/identity/_credentials/on_behalf_of.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    <https://docs.microsoft.com/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow>`_ for a more detailed
+    description of the on-behalf-of flow.
+
+    :param str tenant_id: ID of the service principal's tenant. Also called its "directory" ID.
+    :param str client_id: the service principal's client ID
+    :param str client_secret: one of the service principal's client secrets
+    :param str user_assertion: the access token the credential will use as the user assertion when requesting
+        on-behalf-of tokens
+
+    :keyword bool allow_multitenant_authentication: when True, enables the credential to acquire tokens from any tenant
```
Comment (xiangyan99) [inline sdk/identity/azure-identity/azure/identity/_credentials/on_behalf_of.py:35]: If allow_multitenant_authentication is true, is tenant_id still required?

File: sdk/identity/azure-identity/azure/identity/aio/_credentials/on_behalf_of.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+from ..._internal import validate_tenant_id
+
+if TYPE_CHECKING:
+    from typing import Any
+    from azure.core.credentials import AccessToken
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class OnBehalfOfCredential:
```
Comment (xiangyan99) [inline sdk/identity/azure-identity/azure/identity/aio/_credentials/on_behalf_of.py:21]: Do we want to inherit from abc.ABC?

File: sdk/identity/azure-identity/azure/identity/_credentials/on_behalf_of.py (hunk 1 above)
Comment (chlowell) [reply sdk/identity/azure-identity/azure/identity/_credentials/on_behalf_of.py:35]: Yes.

File: sdk/identity/azure-identity/azure/identity/aio/_credentials/on_behalf_of.py (hunk 2 above)
Comment (xiangyan99) [reply sdk/identity/azure-identity/azure/identity/aio/_credentials/on_behalf_of.py:21]: Sorry. I meant AsyncContextManager. I need some coffee. :)

**Review by xiangyan99 — APPROVED** [review]

File: sdk/identity/azure-identity/azure/identity/aio/_credentials/on_behalf_of.py (hunk 2 above)
Comment (chlowell) [reply sdk/identity/azure-identity/azure/identity/aio/_credentials/on_behalf_of.py]: Right you are. I needed coffee when I wrote this 💤

**Review by mccoyp — APPROVED** [review]

**Pull request merged**